cg/meta/upload/gisaid/gisaid.py

```diff
@@ -28,127 +36,260 @@ def __init__(self, config: CGConfig):
+    ) -> List[List[str]]:
+        """Get completion file with accession numbers"""
+
+        new_completion_file_data = [["provnummer", "urvalskriterium", "GISAID_accession"]]
+        with open(str(completion_file.absolute()), "r") as file:
+            completion_file_reader = csv.DictReader(file, delimiter=",")
+            for sample in completion_file_reader:
+                sample_id = sample["provnummer"]
+                selection_criteria = sample["urvalskriterium"]
+                completion_nr = completion_data.get(sample_id, "")
+                new_completion_file_data.append([sample_id, selection_criteria, completion_nr])
+        return new_completion_file_data
+
+    def update_completion_file(
+        self, completion_file: Path, new_completion_file_data: List[List[str]]
+    ) -> None:
+        """Update completion file with accession numbers"""
+
+        with open(str(completion_file.absolute()), "w", newline="\n") as file:
+            LOG.info("writing accession numbers to new completion file")
+            writer = csv.writer(file)
+            writer.writerows(new_completion_file_data)
+
+    def ensure_log_in_housekeeper(self, case_id: str, file_path: Path, tags: List[str]) -> None:
+        """Adding file to housekeeper if it doesnt exist"""
+
+        if not self.housekeeper_api.find_file_in_latest_version(case_id=case_id, tags=tags):
+            self.housekeeper_api.add_and_include_file_to_latest_version(
+                case_id=case_id, file=file_path, tags=tags
+            )
+
+    def upload(self, case_id: str) -> None:
+        """Uploading results to gisaid and saving the accession numbers in completion file"""
+
+        gisaid_samples: List[GisaidSample] = self.get_gisaid_samples(case_id=case_id)
+        files: UploadFiles = UploadFiles(
+            csv_file=self.build_gisaid_csv(
+                gisaid_samples=gisaid_samples, file_name=f"{case_id}.csv"
+            ),
+            fasta_file=self.build_gisaid_fasta(
+                gisaid_samples=gisaid_samples, file_name=f"{case_id}.fasta", case_id=case_id
+            ),
+            log_file=self.get_gisaid_log_file_path(case_id=case_id),
+        )
+
+        accession_numbers: Dict[str, str] = self.get_accession_numbers(log_file=files.log_file)
+        if len(accession_numbers) == len(gisaid_samples):
+            raise GisaidUploadFailedError(
+                message=f"The samples in bundle {case_id} are already uploaded to GISAID and can not be uploaded more "
+                f"than once. "
+            )
+
+        try:
+            self.upload_results_to_gisaid(files)
+        except Exception as e:
+            raise GisaidUploadFailedError(message=str(e))
+
+        files.csv_file.unlink()
+        files.fasta_file.unlink()
+
+        self.ensure_log_in_housekeeper(
+            case_id=case_id, tags=["gisaid-log"], file_path=files.log_file
+        )
+
+        completion_file: Path = self.housekeeper_api.find_file_in_latest_version(
```

Comment:
this doesnt need to be retrieved unless you have accession numbers

Reply:
:) true